lib/tree.rb

```diff
@@ -16,51 +16,155 @@ def initialize
+      end
+    end
+  end
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
+  def find_helper(current,key)
+    return nil if current.nil?
+    if key == current.key
+      return current.value
+    end
+    if key < current.key 
+      return find_helper(current.left, key)
+    end
+    if key > current.key
+      return find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root,key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n) - an new array is created to store the in-order nodes
+  
   def inorder
-    raise NotImplementedError
+    inorder_array = [] 
+    return inorder_array if @root == nil
+    current = @root
+    return inorder_helper(current, inorder_array)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, inorder_array)  
+    if current == nil
+      return inorder_array
+    else
+      inorder_helper(current.left, inorder_array)
+      inorder_array.push({:key => current.key, :value => current.value})
+      inorder_helper(current.right, inorder_array)
+    end
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)    
   def preorder
-    raise NotImplementedError
+    array = []
+    return array if @root == nil
+    current = @root
+    return preorder_helper(current, array)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current, array)
+    if current == nil
+      return array
+    else
+    array.push({:key => current.key, :value => current.value})
+    preorder_helper(current.left, array)
+    preorder_helper(current.right, array)
+    end
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n) 
   def postorder
-    raise NotImplementedError
+    array = []
+    return array if @root == nil
+    current = @root
+    return postorder_helper(current, array)
+  end
+
+  def postorder_helper(current, array)
+    if current == nil
+      return array
+    else
+    postorder_helper(current.left, array)
+    postorder_helper(current.right, array)
+    array.push({:key => current.key, :value => current.value})
+    end
   end
 
   # Time Complexity: 
   # Space Complexity: 
   def height
-    raise NotImplementedError
+    current = @root
+    height = 1
+    return 0 if @root.nil?
+    return height_helper(current,height)
+  end
+
+  def height_helper(current, height)
+    if current.nil?
+      return height
+    else
+      height += 1
+      height_helper(current.left,height)
+    end
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def bfs
-    raise NotImplementedError
+    queue = Queue.new
+    if @root.nil?
+      return []
+    end
+    array = []
+    queue.push(@root)
+    while !queue.empty? 
+      node = queue.pop
```

Comment:
If you're doing pushes and pops, this isn't a Queue it's a stack.  Thus This isn't going to be doing BFS